plugins/org.eclipse.n4js/src/org/eclipse/n4js/workspace/N4JSProjectConfigSnapshot.java

```diff
@@ -271,4 +283,125 @@ public ImmutableList<ModuleFilterSpecifier> getModuleFilterSpecifiersByType(Modu
 		}
 		return result.build();
 	}
+
+	/** Returns all Uris that are contained in a source folder of this project */
+	public Set<URI> getAllContents(IFileSystemScanner scanner) {
+		return Sets.newLinkedHashSet(
+				IterableExtensions.flatMap(getSourceFolders(), (srcFolder) -> srcFolder.getAllResources(scanner)));
+	}
+
+	/**
+	 * Returns {@code true} iff this is a plain js project in scope '@types'.
+	 * <p>
+	 * If {@code true}, the builder will build only the closure of all included files started from those files defined
+	 * in the tsconfig.json properties {@code files}, {@code include} and in the package.json properties {@code main},
+	 * {@code type}. As a consequence, there might be files in the source folders of this project that remain unbuilt.
+	 * <p>
+	 * If {@code false}, the builder will build all files included in the source folders.
+	 */
+	public boolean hasTsConfigBuildSemantic() {
+		ProjectDescription pd = getProjectDescription();
+		if (pd.getProjectType() != ProjectType.PLAINJS) {
+			return false;
+		}
+		if (pd.getN4JSProjectName() != null && pd.getN4JSProjectName().isScopeTypes()) {
+			return true;
```

Comment:
What about PLAINJS projects that have a `tsconfig.json` file but do not use the `@types` scope?
(I think these days many libraries ship .d.ts files right next to the .js files in a single npm package without providing a separate `@types` package.)

Reply:
Yeah, you're right. I think that must be supported but I was hesitant right now. Maybe this check is only a few days away from being changed again. At the moment, only @types projects have "closure build" semantics but yes, this should be extended to all plain-js projects, at some day (maybe sooner than later!).